### PR TITLE
Parsing partial MP4 containers

### PR DESCRIPTION
**Pull request by roznawsk**
* Parsing partial MP4 containers
* Separated `SampleTableBox` logic from `TrackBox`
* Added functions for unpacking `trak` containers, i.e. operation inverse to `assemble`ing
* Added test for `MP4.Demuxer.ISOM`

### CODE REVIEW

File: test/membrane_mp4/container_test.exs [hunk 1]
```diff
@@ -5,41 +5,53 @@ defmodule Membrane.MP4.ContainerTest do
...
+    data = fixtures |> Path.join(reference) |> File.read!()
+    assert {boxes, <<>>} = data |> Container.parse!()
+    assert boxes |> Container.serialize!() == data
+  end
+
+  defp test_partial(file, boxes_expected) do
+    data = @isom_fixtures |> Path.join(file) |> File.read!()
+    data_size = byte_size(data) - 1
+    <<data::binary-size(data_size), _last::binary-size(1)>> = data
+    assert {:ok, boxes, <<_rest::binary>>} = data |> Container.parse()
```
Comment (balins) [inline test/membrane_mp4/container_test.exs:18]:
```suggestion
    <<partial::binary-size(data_size), _last::binary-size(1)>> = data
    assert {:ok, boxes, <<_rest::binary>>} = Container.parse(partial)
```

I think we should use other variable name after stripping the last byte

File: lib/membrane_mp4/container.ex [hunk 2]
```diff
@@ -26,38 +26,38 @@ defmodule Membrane.MP4.Container do
   @doc """
   Parses binary data to MP4 according to `#{inspect(Schema)}.schema/0`.
   """
-  @spec parse(binary) :: {:ok, t} | {:error, parse_error_context_t}
+  @spec parse(binary) :: {:ok, t, binary()} | {:error, parse_error_context_t}
```
Comment (balins) [inline lib/membrane_mp4/container.ex:29]:
```suggestion
  @spec parse(binary) :: {:ok, t, leftover :: binary()} | {:error, parse_error_context_t}
```

I'd add some explanation what the returned binary is either by using a named type or in `@doc`. Same for `parse/2`

**Review by balins — APPROVED** [review]
Only nitpicks ¯\_(ツ)_/¯

File: test/membrane_mp4/demuxer/isom/demuxer_test.exs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  defp out_path_for(filename), do: "/tmp/out_#{filename}"
+  defp ref_path_for(filename), do: "test/fixtures/payloaded_#{filename}"
+
+  defp prepare_test(filename) do
+    out_path = out_path_for(filename)
+    File.rm(out_path)
+    on_exit(fn -> File.rm(out_path) end)
+  end
```
[the remark this answers is not in the document]
Comment (shuntrho) [reply test/membrane_mp4/demuxer/isom/demuxer_test.exs:24]: Thanks, that looks great!

File: lib/membrane_mp4/track.ex [hunk 4]
```diff
@@ -32,7 +32,13 @@ defmodule Membrane.MP4.Track do
           width: non_neg_integer,
           timescale: pos_integer
         }) :: __MODULE__.t()
-  def new(config) do
+  def new(%{width: width, height: height, content: content, timescale: timescale} = config) do
```
Comment (mat-hek) [inline lib/membrane_mp4/track.ex:35]:
```suggestion
  def new(config) do
    %{width: width, height: height, content: content, timescale: timescale} = config
```

File: lib/membrane_mp4/container/header.ex [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  The `content_size` field is equal to the box size minus the size of the header (8 bytes).
+  """
+
+  @enforce_keys [:type, :content_size]
+
+  defstruct @enforce_keys
+
+  @type t :: %__MODULE__{
+          type: atom(),
```
Comment (mat-hek) [inline lib/membrane_mp4/container/header.ex:13]: I think we're using `name` everywhere

File: lib/membrane_mp4/movie_box/sample_table_box.ex [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+  end
+
+  defp unpack_content(:avc1, %{children: boxes}) do
+    %AVC1{avcc: boxes[:avcC].content, inband_parameters?: false}
+  end
+
+  defp unpack_content(:mp4a, %{children: boxes, fields: fields}) do
+    %AAC{
+      esds: boxes[:esds].fields.elementary_stream_descriptor,
+      sample_rate: fields.sample_rate |> elem(0),
```
Comment (mat-hek) [inline lib/membrane_mp4/movie_box/sample_table_box.ex:230]: don't use `elem`

File: lib/membrane_mp4/movie_box/track_box.ex [hunk 7]
```diff
@@ -158,187 +156,22 @@ defmodule Membrane.MP4.MovieBox.TrackBox do
...
+    header = boxes[:tkhd].fields
+    media = boxes[:mdia].children
+
+    sample_table = SampleTableBox.unpack(media[:minf].children[:stbl])
+
+    %Track{
+      id: header.track_id,
+      content: sample_table.sample_description.content,
+      height: header.height |> elem(0),
+      width: header.width |> elem(0),
```
Comment (mat-hek) [inline lib/membrane_mp4/movie_box/track_box.ex:170]: don't use `elem`

File: test/membrane_mp4/demuxer/isom/demuxer_test.exs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    assert_end_of_stream(pid, :sink, :input)
+    refute_sink_buffer(pid, :sink, _buffer, 0)
+
+    assert :ok == Pipeline.terminate(pid, blocking?: true)
+
+    assert_files_equal(out_path, ref_path)
+  end
+
+  describe "Demuxer.ISOM should demux" do
+    @tag :skip
```
Comment (mat-hek) [inline test/membrane_mp4/demuxer/isom/demuxer_test.exs:31]:
```suggestion
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: test/membrane_mp4/demuxer/isom/demuxer_test.exs (hunk 8 above)
Comment (shuntrho) [reply test/membrane_mp4/demuxer/isom/demuxer_test.exs:31]: Well, the demuxer barely exists now, so the tests will fail.

File: lib/membrane_mp4/container.ex [hunk 9]
```diff
@@ -25,39 +25,46 @@ defmodule Membrane.MP4.Container do
 
   @doc """
   Parses binary data to MP4 according to `#{inspect(Schema)}.schema/0`.
+
+  Returns boxes that have been correctly parsed, and the leftover data that
+  doesn't constitute to an entire box.
```
Comment (balins) [inline lib/membrane_mp4/container.ex:30]:
```suggestion
  is insufficient to constitute a complete box.
```

File: lib/membrane_mp4/container.ex [hunk 10]
```diff
@@ -25,39 +25,46 @@ defmodule Membrane.MP4.Container do
...
+  @spec parse(binary) :: {:ok, t, leftover :: binary()} | {:error, parse_error_context_t}
   def parse(data) do
     parse(data, @schema)
   end
 
   @doc """
   Parses binary data to MP4 according to a custom schema.
+
+  Returns boxes that have been correctly parsed, and the leftover data that
+  doesn't constitute to an entire box.
```
Comment (balins) [inline lib/membrane_mp4/container.ex:41]:
```suggestion
  is insufficient to constitute a complete box.
```

File: lib/membrane_mp4/container/header.ex [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  @header_size @name_size + @size_size
+
+  @doc """
+  Parses the box's header.
+
+  Returns the `t:t/0` and the leftover data.
+  """
+  @spec parse(binary()) :: {:ok, t, leftover :: binary()} | {:error, :not_enough_data}
+  def parse(data) do
+    with <<size::integer-size(@size_size)-unit(8), name::binary-size(@name_size), rest::binary>> <-
```
Comment (balins) [inline lib/membrane_mp4/container/header.ex:28]: I think pattern matching in the function head would be more elegant

File: lib/membrane_mp4/container/header.ex [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+          name: atom(),
+          content_size: non_neg_integer()
+        }
+
+  @name_size 4
+  @size_size 4
+  @header_size @name_size + @size_size
+
+  @doc """
+  Parses the box's header.
```
Comment (balins) [inline lib/membrane_mp4/container/header.ex:22]:
```suggestion
  Parses the header of a box.
```
It's hard to read, even in silence 😆

**Review by balins — CHANGES_REQUESTED** [review]

File: test/membrane_mp4/demuxer/isom/demuxer_test.exs (hunk 8 above)
Comment (mat-hek) [reply test/membrane_mp4/demuxer/isom/demuxer_test.exs:31]: ok, thought it's a leftover

File: lib/membrane_mp4/container/header.ex (hunk 12 above)
Comment (shuntrho) [reply lib/membrane_mp4/container/header.ex:22]: xd

**Review by mat-hek — APPROVED** [review]

**Pull request merged**